mmdet/models/dense_heads/ddod_head.py

```diff
@@ -0,0 +1,760 @@
+                self.loss_single,
+                anchor_list,
+                cls_scores,
+                bbox_preds,
+                iou_preds,
+                # centernesses,
+                labels_list,
+                label_weights_list,
+                bbox_targets_list,
+                bbox_weights_list,
+                reweight_factor_per_level,
+                num_total_samples=num_total_samples)
+
+        anchor_list, valid_flag_list = self.get_anchors(
+            featmap_sizes, img_metas, device=device)
+        cls_reg_targets = self.get_targets(
+            anchor_list,
+            valid_flag_list,
+            cls_scores,
+            bbox_preds,
+            gt_bboxes,
+            img_metas,
+            gt_bboxes_ignore_list=gt_bboxes_ignore,
+            gt_labels_list=gt_labels,
+            label_channels=label_channels,
+            is_cls=False)
+        if cls_reg_targets is None:
+            return None
+
+        (anchor_list, labels_list, label_weights_list, bbox_targets_list,
+         bbox_weights_list, num_total_pos, num_total_neg) = cls_reg_targets
+
+        num_total_samples = reduce_mean(
+            torch.tensor(num_total_pos, dtype=torch.float,
+                         device=device)).item()
+        num_total_samples = max(num_total_samples, 1.0)
+
+        # get pos samples for each level
+        bg_class_ind = self.num_classes
+        for ii, each_level_label in enumerate(labels_list):
+            pos_inds = ((each_level_label >= 0)
+                        & (each_level_label < bg_class_ind)).nonzero().squeeze(1)
+            # num_pos_samples_per_level.append(len(pos_inds))
+            self.reg_num_pos_samples_per_level[ii] += len(pos_inds)
+        # get reweight factor from 1 ~ 2 with bilinear interpolation
+        min_pos_samples = min(self.reg_num_pos_samples_per_level)
+        max_pos_samples = max(self.reg_num_pos_samples_per_level)
+        interval = 1. / (max_pos_samples - min_pos_samples + 1e-10)
+        reweight_factor_per_level = []
+        for pos_samples in self.reg_num_pos_samples_per_level:
+            factor = 2. - (pos_samples - min_pos_samples) * interval
+            reweight_factor_per_level.append(factor)
+
+        reg_losses_cls, reg_losses_bbox, reg_losses_iou = multi_apply(
+                self.loss_single,
+                anchor_list,
+                cls_scores,
+                bbox_preds,
+                iou_preds,
+                # centernesses,
+                labels_list,
+                label_weights_list,
+                bbox_targets_list,
+                bbox_weights_list,
+                reweight_factor_per_level,
+                num_total_samples=num_total_samples)
+
+        return dict(
+            loss_cls=cls_losses_cls,
+            loss_bbox=reg_losses_bbox,
+            loss_iou=reg_losses_iou)
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'iou_preds'))
+    def get_bboxes(self,
```

Comment:
It should be possible to reuse `base_dense_head` without overriding `get_bboxes`.

Reply:
There are a little difference between `base_dense_head.get_bboxes()` and `ddod_head.get_bboxes()`. Therefore, this function will be maintained.

Reply:
Looks the same, can you elaborate on the inconsistency?

Reply:
I have rechecked the function. The input param`score_factors` in `base_dense_head.get_bboxes()` is as same as `iou_preds` in `ddod_head.get_bboxes()`. I have inherited `base_dense_head` to reuse get_bboxes.